examples/tut-6-rpc/client-py/index.py

```diff
@@ -0,0 +1,130 @@
+                break
+        except OSError as ex:
+            time.sleep(0.01)
+            if time.perf_counter() - start_time >= timeout:
+                raise TimeoutError('Waited too long for the port {} '
+                                   'on host {} to start accepting connections.'
+                                   .format(port, host)) from ex
+
+
+def mk_rpc(proto='http',
+           host=os.environ['REACH_RPC_SERVER'],
+           port=os.environ['REACH_RPC_PORT']):
+
+    def rpc(m, *args):
+        wait_for_port(port, host)
+        lab = 'RPC %s %s' % (m, json.dumps([*args]))
+        print(lab)
+        ans = requests.post('%s://%s:%s%s' % (proto, host, port, m),
+                            json=[*args])
+        print('%s ==> %s' % (lab, json.dumps(ans.json())))
+        return ans.json()
+
+    def rpc_callbacks(m, arg, cbacks):
+        vals  = {k: v    for k, v in cbacks.items() if not callable(v)}
+        meths = {k: True for k, v in cbacks.items() if     callable(v)}
+        p     = rpc(m, arg, vals, meths)
+
+        while True:
+            if p['t'] == 'Done':
+                return p
+
+            elif p['t'] == 'Kont':
+                cback = cbacks[p['m']]
+                ans   = cback(*p['args'])
+                p     = rpc('/kont', p['kid'], ans)
+
+            else:
+                # TODO fix swallowed exceptions
+                raise Exception('Illegal callback return: %s' % json.dumps(p))
+
+    return rpc, rpc_callbacks
+
+
+def main():
+    print('I am the client')
+
+    rpc, rpc_callbacks = mk_rpc()
+
+    starting_balance = rpc('/stdlib/parseCurrency', 10)
+    acc_alice        = rpc('/stdlib/newTestAccount', starting_balance)
+    acc_bob          = rpc('/stdlib/newTestAccount', starting_balance)
+
+    fmt          = lambda x: rpc('/stdlib/formatCurrency', x, 4)
+    get_balance  = lambda w: fmt(rpc('/stdlib/balanceOf', w))
+
+    before_alice = get_balance(acc_alice)
+    before_bob   = get_balance(acc_bob)
+
+    ctc_alice    = rpc('/acc/deploy', acc_alice)
+    ctc_bob      = rpc('/acc/attach', acc_bob, rpc('/ctc/getInfo', ctc_alice))
+
+    HAND         = ['Rock', 'Paper', 'Scissors']
+    OUTCOME      = ['Bob wins', 'Draw', 'Alice wins']
+
+    def getHand(who):
+        hand = random.randint(0, 2)  # TODO better source of randomness(?)
+        print('%s played %s' % (who, HAND[hand]))
+        return hand
+
+    acceptWager = lambda amt: print('Bob accepts the wager of %s' % fmt(amt))
+
+    player = lambda who: \
+        {'stdlib.hasRandom': True,
+         'getHand':       lambda:   getHand(who),
+         'informTimeout': lambda:   print('%s observed a timeout' % who),
+         'seeOutcome':    lambda n: print('%s saw outcome %s'
+            % (who, OUTCOME[rpc('/stdlib/bigNumberToNumber', n)]))
+         }
+
+    play_alice = lambda: rpc_callbacks(
+        '/backend/Alice',
+        ctc_alice,
+        dict(wager=rpc('/stdlib/parseCurrency', 5), **player('Alice')))
```

Comment:
Right here you are passing a dictionary to `rpc_callback`. Would it more "Pythonic" to pass an object where the interaction interface methods are methods of the object? Is the code that you already have in `rpc_callbacks` compatible with it being an object?

Reply:
I'll leave this unresolved in case you'd still like to consider the object route (Slack convo provides more detail), but I've replaced the `lambda`s in 4183771.